Simple distributed optimizer

aazzolini opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29304 Simple distributed optimizer**

Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18354586](https://our.internmc.facebook.com/intern/diff/D18354586/)

aazzolini committed:
Simple distributed optimizer

Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18354586](https://our.internmc.facebook.com/intern/diff/D18354586/)

[ghstack-poisoned]

aazzolini commented:
This PR replaces  https://github.com/pytorch/pytorch/pull/28910 which was closed by mistake and couldn't be reopened.

For this version, I got rid of FunctionalOptimizer and made torch.optim.Optimizer work directly with DistributedOptimizer, simplifying the implementation and reducing new API footprint.

zzzwen commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,90 @@
+        kwargs: arguments to pass to the optimizer constructor on each worker.
+    """
+    def __init__(self, optimizer_class, params_rref, *args, **kwargs):
+        per_worker_params_rref = defaultdict(lambda: [])
```
nit: `defaultdict(list)`

pietern reviewed (commented):
Nice! This is very simple indeed :)

vincentqb commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,90 @@
+            [rref.local_value().wait() for rref in local_params_rref],
+            *args,
+            **kwargs)
+        self.lock = Lock()
```
This is the lock mentioned [here](https://github.com/pytorch/pytorch/pull/28910#discussion_r341277257) to, right? How would we modify this to get Hogwild?

aazzolini replied:
Yes. There's no simple way of modifying the current implementation to get hogwild. We'd have to either 1) change the interface (FunctionalOptimizer); or some alternative such as keeping an object pool / or thread local instances of optim.Optimizers to avoid gradient sharing across threads. I can write a comment on why the lock is there.

vincentqb replied:
Cool, just wanted to verify. Thanks!

pritamdamania87 replied:
Wouldn't removing this lock give us some form of hogwild? I'm not sure how accurate this example is: https://pytorch.org/docs/stable/notes/multiprocessing.html#hogwild, although it does something similar where accumulating gradients on the tensors and running the optimizer could interleave among different processes.

albanD commented:
What is the plan for adding a tutorial / example?
Are you planning on adding this here? Or wait to have all the features to do a single tutorial with a full training?

mrshenli commented:
> What is the plan for adding a tutorial / example?

We need thorough docstrings for this one, then we can have a single tutorial for full training I think.

mrshenli commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,90 @@
+        with self.lock:
+            for param, grad in all_local_grads.items():
+                param.grad = grad
+            self.optim.step()
```
this is nice!

mrshenli commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,90 @@
+        params_rref (list[RRef]): list of RRefs to local or remote parameters
+            to optimize.
+        args: arguments to pass to the optimizer constructor on each worker.
+        kwargs: arguments to pass to the optimizer constructor on each worker.
```
Shall we add an example here?

aazzolini replied:
I think that a convincing example will need a way to 1) create the remote module; 2) call it; 3) get a list of RRef params for it. We can add this after we have closed on a way of doing these.

mrshenli commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,90 @@
+    specific parameters.
+
+    Args:
+        optimizer_class (FunctionalOptimizer): the class of optimizer to
```
There is no `FunctionalOptimizer` any more, do you mean `optim.Optimizer`?

mrshenli commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,90 @@
+            to optimize.
+        args: arguments to pass to the optimizer constructor on each worker.
+        kwargs: arguments to pass to the optimizer constructor on each worker.
+    """
```
Do we want to add a warning that we only make sure concurrent `step()` calls will not modify the same param at the same time, but they could still modify params on different owners in an interleaving way? This means that when an dist optimizer tries to apply a grad x to a param, the param might already be different from when grad x was computed. This behavior is by design. If the application needs to do global exclusive dist optimizer `step()`, they will have to synchronize it on their own.

mrshenli reviewed (approved):
This LGTM! My comments are mostly on docs.

aazzolini committed:
Update on "Simple distributed optimizer"


Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18354586](https://our.internmc.facebook.com/intern/diff/D18354586/)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
Internal test failure is real

aazzolini committed:
Update on "Simple distributed optimizer"


Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18354586](https://our.internmc.facebook.com/intern/diff/D18354586/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Test all passed, let's land this!
